modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java

```diff
@@ -0,0 +1,449 @@
+            throw new IllegalArgumentException("Invalid endpoint [" + endpointPattern + "]");
+        }
+        if (endpointPattern.endsWith("/") == false) {
+            endpointPattern += "/";
+        }
+        endpointPattern += "%s/download?suffix=%s";
+
+        // at this point the pattern looks like this (in the default case):
+        // https://download.maxmind.com/geoip/databases/%s/download?suffix=%s
+
+        return Strings.format(endpointPattern, name, suffix);
+    }
+
+    static final String DATABASES_INDEX = ".geoip_databases";
+    static final int MAX_CHUNK_SIZE = 1024 * 1024;
+
+    private final Client client;
+    private final HttpClient httpClient;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+
+    // visible for testing
+    protected volatile EnterpriseGeoIpTaskState state;
+    private volatile Scheduler.ScheduledCancellable scheduled;
+    private final Supplier<TimeValue> pollIntervalSupplier;
+    private final Function<String, HttpClient.PasswordAuthenticationHolder> credentialsBuilder;
+
+    EnterpriseGeoIpDownloader(
+        Client client,
+        HttpClient httpClient,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        long id,
+        String type,
+        String action,
+        String description,
+        TaskId parentTask,
+        Map<String, String> headers,
+        Supplier<TimeValue> pollIntervalSupplier,
+        Function<String, HttpClient.PasswordAuthenticationHolder> credentialsBuilder
+    ) {
+        super(id, type, action, description, parentTask, headers);
+        this.client = client;
+        this.httpClient = httpClient;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.pollIntervalSupplier = pollIntervalSupplier;
+        this.credentialsBuilder = credentialsBuilder;
+    }
+
+    void setState(EnterpriseGeoIpTaskState state) {
+        // this is for injecting the state in GeoIpDownloaderTaskExecutor#nodeOperation just after the task instance has been created
+        // by the PersistentTasksNodeService -- since the GeoIpDownloader is newly created, the state will be null, and the passed-in
+        // state cannot be null
+        assert this.state == null;
+        assert state != null;
+        this.state = state;
+    }
+
+    // visible for testing
+    void updateDatabases() throws IOException {
+        var clusterState = clusterService.state();
+        var geoipIndex = clusterState.getMetadata().getIndicesLookup().get(EnterpriseGeoIpDownloader.DATABASES_INDEX);
+        if (geoipIndex != null) {
+            logger.trace("The {} index is not null", EnterpriseGeoIpDownloader.DATABASES_INDEX);
+            if (clusterState.getRoutingTable().index(geoipIndex.getWriteIndex()).allPrimaryShardsActive() == false) {
+                throw new ElasticsearchException("not all primary shards of [" + DATABASES_INDEX + "] index are active");
```

Comment:
Should this be an actual exception (with the error logging this would cause), or should it be more of a warning/ignorable thing? I'm worried about spitting a bunch of errors and logging whenever a node restarts.

Reply:
This was copy/pasted from GeoIpDownloader (this change added in https://github.com/elastic/elasticsearch/pull/86842). We were attempting to keep them pretty similar so that they could be merged back into one thing in the future. This seems like a small enough difference though that I could just change it to a debug level message.